python-threatexchange/threatexchange/extensions/vpdq/vpdq_index.py

```diff
@@ -0,0 +1,100 @@
+# Copyright (c) Facebook, Inc. and its affiliates. All Rights Reserved
+
+"""
+Implementation of SignalTypeIndex abstraction for PDQ by wrapping
+hashing.pdq_faiss_matcher.
+"""
+
+import typing as t
+import vpdq
+from threatexchange.signal_type.index import (
+    PickledSignalTypeIndex,
+    VPDQIndexMatch,
+    IndexMatch,
+    T as IndexT,
+)
+from threatexchange.extensions.vpdq.vpdq_faiss import VPDQHashIndex
+from threatexchange.extensions.vpdq.vpdq_util import (
+    prepare_vpdq_feature,
+    VPDQ_QUALITY_THRESHOLD,
+    VPDQ_DISTANCE_THRESHOLD,
+)
+
+
+class VPDQIndex(PickledSignalTypeIndex[IndexT]):
+    """
+    Wrapper around the vpdq faiss index lib using VPDQHashIndex
+    """
+
+    def __init__(
+        self,
+        quality_threshold: int = VPDQ_QUALITY_THRESHOLD,
+    ) -> None:
+        super().__init__()
+        self.index: VPDQHashIndex = VPDQHashIndex()
+        self._entry_idx_to_features_and_entires: t.List[
+            t.Tuple[t.List[vpdq.VpdqFeature], IndexT]
+        ] = []
+        self._index_idx_to_vpdqHex_and_entry: t.List[t.Tuple[int, t.List[int]]] = []
+        self._unique_vpdqHex_to_idx: t.Dict[str, int] = {}
+        self.quality_threshold = quality_threshold
+
+    def add(self, signal_str: str, entry: IndexT) -> None:
+        entry_id = len(self._entry_idx_to_features_and_entires)
+        features = prepare_vpdq_feature(signal_str, self.quality_threshold)
+        self._entry_idx_to_features_and_entires.append((features, entry))
+        # Use hex to represent the feature because it saves the space
```

Comment:
There are even more compact serialization, such as bytes.

Reply:
I think I will leave it as it is for easier checking if the feature(hex) already exists in _unique_vpdqHex_to_idx.